Week0 - mega PR!

nowireless opened the pull request:
the week0 branch has all stuff we have been working on, and maybe unstable!

abrtn committed:
Add constants for trajectory finding

abrtn committed:
Add constants for trajectory finding

abrtn committed:
Insert new drive subsystem

abrtn committed:
Insert new drive subsystem

abrtn committed:
Add integrated encoders to characterization

abrtn committed:
Added integrated encoders to characterization

abrtn committed:
Added integrated encoders to characterization

abrtn committed:
Add Ramsete and default trajectory

abrtn committed:
Add Ramsete and default trajectory

abrtn committed:
Merge remote-tracking branch 'origin/feature/add-characterization' into feature/add-characterization

abrtn committed:
Add gyro to TalonSRX

abrtn committed:
Add gyro to TalonSRX

abrtn committed:
Merge remote-tracking branch 'origin/main' into feature/add-characterization

nowireless committed:
Created encoderCounts and encoderRateToMetersPerSecond helper functions

Start using Units class instead of magic constants.
Found that the encoder rate functions where using sensor position instead of rate.

nowireless committed:
Created encoderCounts and encoderRateToMetersPerSecond helper functions

Start using Units class instead of magic constants.
Found that the encoder rate functions where using sensor position instead of rate.

nowireless committed:
Add a unit tests for encodeCountsToMeters and encoderRateToMetersPerSecond.

nowireless committed:
Add a unit tests for encodeCountsToMeters and encoderRateToMetersPerSecond.

shetaye committed:
Merge pull request #13 from Team2470/feature/drive-unittests

Refactored how how encoder counts are converted into rate and distance, and added unit tests.

shetaye committed:
Use method references

I did not know these existed, and they make for much cleaner code.

shetaye committed:
Use method references

I did not know these existed, and they make for much cleaner code.

shetaye committed:
Break example code into dedicated files

Object instances that should remain constant were moved to Constants and RamseteCommand was subclassed as DriveAlongTrajectoryCommand to provide a nicer interface.

shetaye committed:
Break example code into dedicated files

Object instances that should remain constant were moved to Constants and RamseteCommand was subclassed as DriveAlongTrajectoryCommand to provide a nicer interface.

abrtn committed:
Add pigeon gyro to config file

abrtn committed:
Add pigeon gyro to config file

Harrisyiwei committed:
Merge remote-tracking branch 'origin/main' into main

Harrisyiwei committed:
Create Vision subsystem

Harrisyiwei committed:
Create Vision subsystem

Harrisyiwei committed:
Add vision code

Harrisyiwei committed:
Add vision code

shetaye committed:
Format code

shetaye committed:
Format code

max-mckinley committed:
Added conveyor code

max-mckinley committed:
spell check

jack12324 committed:
Merge branch 'main' into feature/conveyer

shetaye committed:
Merge remote-tracking branch 'origin/feature/add-characterization' into feature/add-characterization

shetaye committed:
Remove duplicate constants

abrtn committed:
Add changes to run conveyor and shooter

Harrisyiwei committed:
Add AutoTest Command

abrtn committed:
Start autonomous

Harrisyiwei committed:
Add new Conveyor Command file

abrtn committed:
Add basic autonomous

nowireless committed:
Invert shooter motor so positive values mean something.

nowireless committed:
REworked auto to be more commandy

nowireless committed:
Merge branch 'feature/conveyer' into week0

nowireless committed:
Merge branch 'feature/AutoTest' into week0

nowireless committed:
Merge remote-tracking branch 'origin/feature/add-characterization' into week0

max-mckinley committed:
Dpad support for climber

max-mckinley committed:
move conveyor buttons

max-mckinley committed:
Assign gyro can id

max-mckinley committed:
start intake subsystem

shetaye committed:
Finish intake subsystem prototype

shetaye committed:
Add intake commands

shetaye committed:
Bind intake to Y

shetaye committed:
Merge branch 'feature/intake' into week0

shetaye committed:
Replace gyro talon with conveyor talon

shetaye committed:
Create climber interface to simplify commands

shetaye committed:
Add ClimbAngleCommand

nowireless commented on src/main/java/frc/robot/subsystems/Intake.java:
```diff
@@ -0,0 +1,38 @@
+
+    public Intake() {
+        m_intakeMotor = new CANSparkMax(Constants.kIntakeMotorId, MotorType.kBrushless);
+        m_deploy = new Solenoid(PneumaticsModuleType.CTREPCM, Constants.kIntakeSolenoid);
```
Change `PneumaticsModuleType.CTREPCM` to `PneumaticsModuleType.REVPH`. Because we are using the [REV Pneumatics hub](https://www.revrobotics.com/rev-11-1852/).

nowireless replied:
We also need to fix the FrontClimber: https://github.com/Team2470/FRC-2022-robot/blob/week0/src/main/java/frc/robot/subsystems/FrontClimber.java#L31

max-mckinley replied:
@KarlWheezer

nowireless commented on src/main/java/frc/robot/commands/DeployIntakeCommand.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    public boolean isFinished() {
+        // TODO: Make this return true when this Command no longer needs to run execute()
```
I would argue that this command would never need to decide for itself to end. It can be interrupted when letting go of a button (ie using button.whenPresses), or via a different event.

Though we maybe could use this if we wanted to have an end condition be when the conveyor is holding 2 balls. Not sure if it makes sense to it here, or at a higher level.

shetaye replied:
The `isFinished` method is autogenerated by the WPIlib plugin.  I'll write a commit that removes these stubs.

nowireless replied:
What is the default for isFinished in command Base? I’m fine with having isFinished return false always, if that is not the default behavior.

shetaye replied:
It returns false by default
<img width="621" alt="Screen Shot 2022-02-23 at 12 04 32 PM" src="https://user-images.githubusercontent.com/12775864/155379870-4882bb2e-ea74-45c3-b795-3116e8c2f056.png">
(screenshot from Command class)

shetaye replied:
I removed as many as I could in 29c999a

nowireless commented on src/main/java/frc/robot/commands/DeployIntakeCommand.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    public void end(boolean interrupted) {
+        this.m_intake.stopIntakeMotor();
```
If we are done deploying the intake, shouldn't we send it back up? Then we don't need a dedicated RetractIntakeCommand. 

We have a solenoid with a spring return, so when we are not activating the solenoid the intake we be defaulting to up. Also if we are turning off the motor we should also just raise the intake up.

nowireless replied:
**This is an idea/opinion, not something we have to do. Both ways are functionally the same**

shetaye replied:
We decided to table this and if it becomes an issue (i.e. if commands get messy) we'll revisit it.

nowireless commented on src/main/java/frc/robot/subsystems/Conveyor.java:
```diff
@@ -0,0 +1,41 @@
+
+  public WPI_TalonSRX getMotor() { return this.m_conveyorMotor; }
+
+  public void move(int direction, double speed) {
```
I think we should have a dedicated up/down methods with a predefined speed values for the conveyor. IMHO I don't think we need to have a method that exposes the motor like this.

awesta00 replied:
b31187199838fb62598f144b0144491128d653d0

nowireless commented on src/main/java/frc/robot/subsystems/Conveyor.java:
```diff
@@ -0,0 +1,41 @@
+    m_conveyorMotor.set(ControlMode.PercentOutput, direction*speed);
+  }
+
+  public void moveAtSpeed (double speed) {
```
Whats the point of this if we have move (or vice versa)? They do effectively the same thing.

awesta00 replied:
b31187199838fb62598f144b0144491128d653d0

nowireless commented on src/main/java/frc/robot/commands/ClimbAngleCommand.java:
```diff
@@ -0,0 +1,48 @@
+    public void execute() {
+        if (getError().getDegrees() > 0) {
+            this.m_climber.startClimbMotor(Constants.kClockwise, Constants.kClimberSpeed);
+        } else if (getError().getDegrees() < 0) {
```
I think we can change `} else if (getError().getDegrees() < 0) {` to `} else {`.

Also what happens if the error is some how 0? I know this is unlikely, but what if it becomes zero before isFInished is checked 🙂

max-mckinley replied:
@KarlWheezer

shetaye replied:
970028589fe372432f6cc6d8c8110f4306ca3576

nowireless commented on src/main/java/frc/robot/subsystems/Intake.java:
```diff
@@ -0,0 +1,38 @@
+        m_deploy.set(false);
+    }
+
+    public void startIntakeMotor(int direction, double speed) {
```
I we should make a function called `spin` or something that takes no arguments. 

This `start*Motor` pattern we have I would argue is slightly breaking the abstraction of what a subsystem should be and a lot of the time we are using speed from constants defined in Constants.java. So we might as well make functions thats abstract that away from the commands.

max-mckinley replied:
@KarlWheezer

shetaye replied:
970028589fe372432f6cc6d8c8110f4306ca3576

nowireless commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -52,27 +50,36 @@ public RobotContainer() {
+    rpmButton3.whileHeld(new RunShooterCommand(m_shooter, 2500));
+    //: Climber control
+    DPadTrigger ForwardClimbClockwiseButton = new DPadTrigger(m_controller, DPadTrigger.DPad.kUp);
+    ForwardClimbClockwiseButton.whileActiveContinuous(new ClimbClockwise(m_frontClimber));
```
What does clockwise/counter clockwise mean? How is it defined? 

Perhaps for the climber we should use terminology like `MoveClimberOutward` or `MoveClimberInward`. The climbers we have are inverted from each other so if we can pick something that is clearer it will help prevent us from causing problems later on if we accidentally have the climber move the wrong way.

shetaye replied:
d9c70d42e73a6290b7082afa2fd2dbf5900a5db6, 8472ab11be2d67c8ad343014e89dc79b56bf0662, and ad29246bd4c204deacd65d797183e3d176f14560

nowireless committed:
Upgrade WPILib and CTRE libraries

max-mckinley committed:
Refactor

abrtn committed:
Merge pull request #22 from Team2470/feature/update-wpilib

Upgrade WPILib and CTRE libraries

abrtn committed:
Change IDs and Intake

shetaye committed:
Remove redundant stubs

The FRC plugin autogenerates stubs that are identical to default behavior.

shetaye committed:
Refactor climbing commands to abstract away directionality

shetaye committed:
Rename buttons to reflect climber changes

shetaye committed:
Rename climber commands

awesta00 committed:
Modified conveyer speed code

abrtn committed:
Tuned CAN bus

abrtn committed:
Merge remote-tracking branch 'origin/week0' into week0

awesta00 committed:
Shooter command foundation

abrtn committed:
Tweak climber

abrtn committed:
Revert "Merge remote-tracking branch 'origin/week0' into week0"

This reverts commit b34afe38962d49df3e9df3be87cd0bd69811e863.

abrtn committed:
Merge remote-tracking branch 'origin/week0' into week0

abrtn committed:
Revert "Revert "Merge remote-tracking branch 'origin/week0' into week0""

This reverts commit 216af26e1b8f08c18017235c8b936b90f0934b4c.

abrtn committed:
Climber Arms Troubleshooting

abrtn committed:
Merge tag '2/23/22Working' into feature/vision

nowireless commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -82,6 +111,22 @@ private void configureButtonBindings() {
     // Return null for no autonomous command
-    return null;
+    return new ParallelCommandGroup(
+            new RunShooterCommand(m_shooter, () -> 2500),
```
This is neat 🙂

abrtn committed:
Vision work and conveyor tweaks

abrtn committed:
Simple 1 ball auto

shetaye committed:
Make suppliers optional

abrtn committed:
Vision work

abrtn committed:
Merge branch 'feature/vision' into week0

shetaye committed:
Rewrite shoot command to use new Vision subsystem

shetaye committed:
Clean code

nowireless commented on src/main/java/frc/robot/commands/WaitForShooterRPMCommand.java:
```diff
@@ -0,0 +1,30 @@
+  /** Creates a new WaitForShooterRPMCommand. */
+  public WaitForShooterRPMCommand(Shooter shooter, IntSupplier rpmSupplier, double tolerance) {
+    // Use addRequirements() here to declare subsystem dependencies.
+    addRequirements(shooter);
```
This is the error we were getting last night when we deployed the the latest code form this branch. This WaitForShooterRPMCommand doesn't need to require the shooter subsystem as it fights with the RunShooterCommand for ownership. You can think of this command more of listening, instead of controlling the subsystem. So therefore we don't need to require the shooter subsystem in this command.

```
Error at frc.robot.commands.ShootCommandGroup.<init>(ShootCommandGroup.java:28): Unhandled exception: java.lang.IllegalArgumentException: Multiple commands in a parallel group cannotrequire the same subsystems
        at edu.wpi.first.wpilibj2.command.ParallelDeadlineGroup.addCommands(ParallelDeadlineGroup.java:69)
        at edu.wpi.first.wpilibj2.command.ParallelDeadlineGroup.<init>(ParallelDeadlineGroup.java:39)
        at frc.robot.commands.ShootCommandGroup.<init>(ShootCommandGroup.java:28)
        at frc.robot.RobotContainer.configureButtonBindings(RobotContainer.java:125)
        at frc.robot.RobotContainer.<init>(RobotContainer.java:55)
        at frc.robot.Robot.robotInit(Robot.java:30)
        at edu.wpi.first.wpilibj.TimedRobot.startCompetition(TimedRobot.java:107)
        at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:373)
        at edu.wpi.first.wpilibj.RobotBase.lambda$startRobot$0(RobotBase.java:443)
        at java.base/java.lang.Thread.run(Thread.java:829)

```

I recreated the error by running our code in simulation.

shetaye replied:
463f5e4d04767b600c479901454f667177027257

shetaye committed:
Remove unnecessary requirements and clean code

Commands should only require subsystems that they act upon, not just read data from.

shetaye committed:
Stub conveyor distance and write wait for shooter

abrtn committed:
Fix climber CANcoder readings

abrtn committed:
Get climber arms and soft limits working

abrtn committed:
Conveyor and auto intake work

abrtn committed:
Add conveyor CANCoder

shetaye committed:
Implement conveyor distance command

shetaye committed:
Reformat code

I used IDEA to reformat code using 2 spaces since I believe everyone else is using 2 spaces but I have been using 4.

IronBlades3267 committed:
Checkin

abrtn committed:
Added drive distance command and changed autonomous

IronBlades3267 committed:
Update climber subsystem and command

shetaye committed:
Add conveyor test commands in shuffleboard

shetaye committed:
Merge branch 'climbing' into week0

shetaye committed:
Fix merge issues with brackets

shetaye committed:
Use conveyor sendable

IronBlades3267 committed:
Update climber work to traverse

nowireless commented on src/main/java/frc/robot/subsystems/BackClimber.java:
```diff
@@ -5,58 +5,109 @@
+    // TODO Auto-generated method stub
+
+    if(output>0.1){
+      output = 0.1;
```
This value should get be increased so we have can have back climber arms to move at a reasonable speed. Right now its kind of slow. We want it slow/fast enough so we can gracefully move the robot into position to get the front arms onto the next bar.

nowireless commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -28,51 +30,115 @@
+    conveyorDown.whileHeld(new RunConveyorCommand(m_conveyor, Direction.kDown));
+    // //: Shooter control
+    // JoystickButton rpmButton1 = new JoystickButton(m_buttopad, 8);
+    // rpmButton1.whileHeld(new RunShooterCommand(m_shooter, () -> 2800));
```
We need to re-add all of these shooter buttons. We are going to add another row of buttons on the button pad.

shetaye replied:
74b25823dc7c4f3f6f0091066ab12013db1cc829

nowireless commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -28,51 +30,115 @@
+    MoveClimbToAngle.whenPressed(new ClimbAngleCommand(m_backClimber, Rotation2d.fromDegrees(25)).perpetually());
+
+    JoystickButton MoveClimbToAngle2 = new JoystickButton(m_buttopad, 6);
+    MoveClimbToAngle2.whenPressed(new ClimbAngleCommand(m_backClimber, Rotation2d.fromDegrees(90)).perpetually());
```
Need to make button 7 move the back climber to 125 degrees

nowireless commented on src/main/java/frc/robot/commands/ClimbAngleCommand.java:
```diff
@@ -0,0 +1,51 @@
+  private final BackClimber m_climber;
+  private final Rotation2d m_setpoint;
+
+  public ClimbAngleCommand(BackClimber climber, Rotation2d setpoint) {
```
We should rename this to BackClimbAngleCommand. 

We should also create a FrontClimbAngleCommand, so we can move the front climb arms to a specified angle. I don't think we need PID for the front, so we could probably go back to this version for a start: https://github.com/Team2470/FRC-2022-robot/blob/e7edc3c911a0c4afca847b528dd9b867eb84be94/src/main/java/frc/robot/commands/ClimbAngleCommand.java

nowireless committed:
Some climber notes

shetaye committed:
Add tentative rpm setter

Harrisyiwei committed:
Week One final tweaks

abrtn committed:
Reorganized autonomous for new shooter angle

shetaye committed:
Rename ClimbAngleCommand to BackClimbAngleCommand

shetaye committed:
Create FrontClimbAngleCommand
